content/posts/2025-02-05-perf-improvements.md

```diff
@@ -0,0 +1,237 @@
+
+But the crux of the matter was that our map is accessed for writes by different hook points. For instance, there is the [TC hook](https://github.com/netobserv/netobserv-ebpf-agent/blob/2c96c420fc5ed223b7b99b00b2705fe84c5c0110/bpf/flows.c#L282) that is our main observation point for capturing packets, and other trace points such as [the one looking for drops](https://github.com/netobserv/netobserv-ebpf-agent/blob/2c96c420fc5ed223b7b99b00b2705fe84c5c0110/bpf/pkt_drops.h#L104), which enriches the flows with additional information. Those enrichment trace points can't acquire the lock to perform updates. Using a per-CPU map has been an answer to this problem, as it is not going to be written concurrently.
+
+The drawback of using per-CPU maps for network flows is that for the same flow, which is defined (roughly — more on that later) by its 5-tuple (source IP, destination IP, source port, destination port, protocol), packets have no reason to be processed by just a single CPU. In other words, the flow information is spread over multiple CPU buckets in the map. Not only does it force us to reassemble flow chunks when they are read from the user space, wasting precious CPU cycles; but most importantly, it's not memory efficient, with map sizes being potentially multiplied by the number of cores, and data being sometimes duplicated on several CPU buckets.
+
+Today, we change all of that by refactoring our data structures. Our solution is to avoid writing the flow map from the enrichment hooks. Instead of that, we are introducing a new map, dedicated to enriched data. Map keys are going to be duplicated across those two maps, but it's a lesser evil. So now, we can change the main map to be a shared one across CPUs, with a spinlock. We still have some reassembling work to do in the user space though, to merge the main map with the enrichment map, but it is more straightforward than merging entire flows together. We also have a couple of ideas to further improve this process, more on that later.
+
+Splitting data between a main map and an enrichment map has another benefit: when no enrichment is needed (e.g. when none of the [agent features](https://github.com/netobserv/network-observability-operator/blob/main/docs/FlowCollector.md#flowcollectorspecagentebpf-1) are enabled), no memory is allocated for them, resulting — again — in a more efficient memory usage.
```

Comment:
this isn't the case because the interface list is in the additional map so it will always be allocated

Reply:
remember, we moved it to the main map: https://github.com/netobserv/netobserv-ebpf-agent/pull/509 :-)

Reply:
true but the map still allocated even with no features I don't think we changed that back ?

Reply:
the map is allocated but with `BPF_F_NO_PREALLOC` so its size depends on its content